### PR TITLE
Add unit tests to test DDP with TensorPipe GPU support.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43940 Add unit tests to test DDP with TensorPipe GPU support.**

DDP + RPC supported combining DDP with RPC but was only limited to
testing CPU coverage for RPC.

With TensorPipe now supporting GPU tensors, I'm adding a test to cover RPC with
GPU combined with DDP.

Differential Revision: [D23444241](https://our.internmc.facebook.com/intern/diff/D23444241/)

### CODE REVIEW

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 1]
```diff
@@ -4137,3 +4142,53 @@ def test_device_maps_remote(self):
...
+        out.sum().backward()
+
+        fc_layer2 = copy.deepcopy(fc_layer1)
+        ddp_model = DistributedDataParallel(fc_layer2.cuda(self.rank), device_ids=[self.rank])
+        dist.barrier()
+
+        # Run distributed model.
+        with dist_autograd.context() as context_id:
+            res = rpc.rpc_sync(dst, torch.add, args=(t, t))
+            rpc.api._wait_all_workers()
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:4188]: Why do we need to wait all workers here? Is it due to the deadlock issue mentioned offline? Should we file an issue describing how the deadlock can happen?

**Review by rohan-varma — APPROVED** [review]
LGTM

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 2]
```diff
@@ -4137,3 +4142,53 @@ def test_device_maps_remote(self):
...
+        out = ddp_model(torch.add(t, t))
+        out.sum().backward()
+
+        fc_layer2 = copy.deepcopy(fc_layer1)
+        ddp_model = DistributedDataParallel(fc_layer2.cuda(self.rank), device_ids=[self.rank])
+        dist.barrier()
+
+        # Run distributed model.
+        with dist_autograd.context() as context_id:
+            res = rpc.rpc_sync(dst, torch.add, args=(t, t))
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/testing/_internal/distributed/rpc/rpc_test.py:4187]: Currently, it looks like we just verify remote gradients for an add() and everything else in DDP is local, but ideally the test should cover calling DDP(model) where `model` has both remote and local layers?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 3]
```diff
@@ -4137,3 +4142,53 @@ def test_device_maps_remote(self):
         )
 
         self.assertEqual(rref.to_here(), torch.ones(2).to(1))
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(4)
+    def test_with_ddp(self):
```
Comment (lw) [inline torch/testing/_internal/distributed/rpc/rpc_test.py:4148]: I'm not familiar with these tests: what is the reason to add this to the `rpc_test` file rather than to the `ddp_under_dist_autograd_test` file?

**Review by lw — COMMENTED** [review]
Should we perhaps wait to land this until we implement the intended stream behavior in the TensorPipe agent (see #44084)?

**Comment by pritamdamania87** [comment]
> Should we perhaps wait to land this until we implement the intended stream behavior in the TensorPipe agent (see #44084)?

I think this makes sense since when I enhanced the current unit test based on Rohan's suggestions, we run into a separate synchronization issue since an RPC and allreduce run on the same device and stream concurrently causing a similar as mentioned before deadlock.
